modules/tinymce/src/core/main/ts/fmt/FormatChanged.ts

```diff
@@ -5,101 +5,171 @@
-    Tools.each(parents, (node: Node) => {
-      if (editor.formatter.matchNode(node, format, {}, data.similar)) {
-        if (formatsList.indexOf(format) === -1) {
-          // Execute callbacks
-          Arr.each(data.callbacks, (callback: FormatChangeCallback) => {
-            callback(true, { node, format, parents });
-          });
-
-          newFormats[format] = data.callbacks;
-        }
+  editor.on('FormatApply FormatRemove', (e: EditorEvent<FormatEvent>) => {
+    const element = Optional.from(e.node)
+      .map((nodeOrRange) => FormatUtils.isNode(nodeOrRange) ? nodeOrRange : nodeOrRange.startContainer)
+      .bind((node) => NodeType.isElement(node) ? Optional.some(node) : Optional.from(node.parentElement))
+      .getOrThunk(() => editor.selection.getStart());
 
-        matchedFormats[format] = data.callbacks;
-        return false;
-      }
+    updateAndFireChangeCallbacks(editor, element, registeredFormatListeners.get());
+  });
+};
 
-      if (MatchFormat.matchesUnInheritedFormatSelector(editor, node, format)) {
+const matchingNode = (editor: Editor, parents: Element[], format: string, similar: boolean, vars?: FormatVars): Optional<Element> => {
+  const isMatchingNode = (node: Element) => {
+    const matchingFormat = editor.formatter.matchNode(node, format, vars ?? {}, similar);
+    return !Type.isUndefined(matchingFormat);
+  };
+  const isUnableToMatch = (node: Element) => {
+    if (MatchFormat.matchesUnInheritedFormatSelector(editor, node, format)) {
+      return true;
+    } else {
+      if (!similar) {
+        // If we want to find an exact match, then finding a similar match halfway up the parents tree is bad
+        return Type.isNonNullable(editor.formatter.matchNode(node, format, vars, true));
+      } else {
         return false;
       }
-    });
-  });
-
-  // Check if current formats still match
-  const remainingFormats = filterRemainingFormats(currentFormats.get(), matchedFormats, elm, parents);
-
-  // Update the current formats
-  currentFormats.set({
-    ...newFormats,
-    ...remainingFormats
-  });
+    }
+  };
+  return Arr.findUntil(parents, isMatchingNode, isUnableToMatch);
 };
 
-const filterRemainingFormats = (currentFormats: FormatCallbacks, matchedFormats: FormatCallbacks, elm: Element, parents: Node[]) => Obj.bifilter(currentFormats, (callbacks: FormatChangeCallback[], format: string) => {
-  if (!Obj.has(matchedFormats, format)) {
-    // Execute callbacks
-    Arr.each(callbacks, (callback: FormatChangeCallback) => {
-      callback(false, { node: elm, format, parents });
-    });
+const getParents = (editor: Editor, elm?: Element): Element[] => {
+  const element = elm ?? editor.selection.getStart();
+  return Arr.filter(FormatUtils.getParents(editor.dom, element), (node) =>
+    NodeType.isElement(node) && !NodeType.isBogus(node)
+  );
+};
 
+const hasChanged = (state: Cell<boolean>, newState: boolean): boolean => {
```

Comment:
I missed this earlier, but I'm not very fond of this mutating state while looking it up (I'd never expect a check function to change the state). Can we move the `state.set` call to where this returns true instead please?